### PR TITLE
Enable regex search

### PR DESCRIPTION
**Pull request by SchrodingersGat**
This PR enables "whole word search" and "regex search" when searching via the API.

Users can enable regex search functionality (for their own results)

image

image

This PR also serves as an entry point into further search result customization

Closes https://github.com/inventree/InvenTree/issues/4510

### CODE REVIEW

File: InvenTree/InvenTree/ready.py [hunk 1]
```diff
@@ -42,6 +41,9 @@ def canAppAccessDatabase(allow_test: bool = False, allow_plugins: bool = False):
         'mediarestore',
     ]
 
+    if not allow_shell:
```
Comment (matmair) [inline InvenTree/InvenTree/ready.py:44]: @SchrodingersGat what has this to do with search?

File: InvenTree/order/api.py [hunk 2]
```diff
@@ -1304,7 +1304,7 @@ def download_queryset(self, queryset, export_format):
 
     filter_backends = [
         rest_filters.DjangoFilterBackend,
-        filters.SearchFilter,
+        InvenTreeSearchFilter,
         filters.OrderingFilter,
     ]
```
Comment (matmair) [inline InvenTree/order/api.py:1309]: Suggestion: Seems like this block is used often - can de deduplicate that?

**Review by matmair — APPROVED** [review]

File: InvenTree/InvenTree/ready.py (hunk 1 above)
Comment (SchrodingersGat) [reply InvenTree/InvenTree/ready.py:44]: Nothing, but during testing for search I found that some functions were blocked entirely if running from the shell.

File: InvenTree/order/api.py (hunk 2 above)
Comment (SchrodingersGat) [reply InvenTree/order/api.py:1309]: Yeah, this would be a good candidate for a refactoring

File: InvenTree/InvenTree/ready.py (hunk 1 above)
Comment (matmair) [reply InvenTree/InvenTree/ready.py:44]: Ah ok, I was a bit surprised what effect it could have on search 👍

**Pull request merged**